test/distributed/org/apache/cassandra/distributed/test/RepairBoundaryTest.java

```diff
@@ -0,0 +1,197 @@
+        assertRows(c2Row(1001, 1999, 2000, 2001, 2999, 3000), cluster.get(3).executeInternal(ALL));
+    }
+
+    /**
+     * Insert on every token boundary and + or - to first replica.
+     */
+    void populate()
+    {
+        try
+        {
+            cluster.schemaChange(withKeyspace("DROP TABLE IF EXISTS %s.test;"));
+            cluster.schemaChange(withKeyspace("CREATE TABLE %s.test (k blob, c1 text, c2 text," +
+                                              "PRIMARY KEY (k))"));
+
+            for (int i = 1000; i <= 3000; i += 1000)
+            {
+                keys.put(i, keyForToken(new LongToken(i)));
+                keys.put(i - 1, keyForToken(new LongToken(i - 1)));
+                keys.put(i + 1, keyForToken(new LongToken(i + 1)));
+
+                cluster.coordinator(1).execute(INSERT, ConsistencyLevel.ALL, keys.get(i), String.valueOf(i));
+                cluster.coordinator(1).execute(INSERT, ConsistencyLevel.ALL, keys.get(i - 1), String.valueOf(i - 1));
+                cluster.coordinator(1).execute(INSERT, ConsistencyLevel.ALL, keys.get(i + 1), String.valueOf(i + 1));
+            }
+        }
+        catch (Throwable t)
+        {
+            cluster.close();
+            throw t;
+        }
+    }
+
+    @Test
+    public void primaryRangeRepair()
+    {
+        populate();
+        verify();
+
+        delete(cluster.get(1), 999, 1000, 3001);
+        delete(cluster.get(2), 1999, 2000, 1001);
+        delete(cluster.get(3), 2999, 3000, 2001);
+
+        cluster.forEach(i -> {
+            i.runOnInstance(() -> {
+                try
+                {
+                    Map<String, String> options = new HashMap<>();
+                    options.put("primaryRange", "true");
+                    options.put("incremental", "false");
+                    SimpleCondition await = new SimpleCondition();
+                    StorageService.instance.repair(KEYSPACE, options, ImmutableList.of((tag, event) -> {
+                        if (event.getType() == ProgressEventType.COMPLETE)
+                            await.signalAll();
+                    })).right.get();
+                    await.await(1L, MINUTES);
+                }
+                catch (Exception e)
+                {
+                }
+            });
+        });
+
+        assertRows(c2Row(), cluster.get(1).executeInternal(ALL));
+        assertRows(c2Row(), cluster.get(2).executeInternal(ALL));
+        assertRows(c2Row(), cluster.get(3).executeInternal(ALL));
+    }
+
+    @Test
+    public void singleTokenRangeRepair()
+    {
+        populate();
+        verify();
+
+        delete(cluster.get(1), 999, 1000);
+        delete(cluster.get(3), 1001);
+
+        cluster.get(2).runOnInstance(() -> {
+            try
+            {
+                Map<String, String> options = new HashMap<>();
+                options.put("ranges", "999:1000");
```

Comment:
will file a jira to document this... way too much conversation on a single line =D